### PR TITLE
Fix epoch=None crash in LR-tuning validation with a loss schedule

### LINKED ISSUE #1275: LR-tuning validation crashes when a loss schedule is active (validation batches carry epoch=None)
## Observed

Training run [beaker 01KTG13T](https://beaker.org/ex/01KTG13T8CJHHDDFX187EGKJ6Y) (config `train-4deg-daily-v1-labels-residual-winds-anomaly-tend-reg-rollout.yaml`, launched from `80b2be8a1` on `experiment/2026-06-05-aimip-like`) crashed at epoch 45 on 2026-06-08 18:05 UTC, and crashes again identically on every resume:

```
  File "/gantry-runtime/fme/core/generics/trainer.py", line 391, in _maybe_tune_lr
    new_lr = run_lr_tuning_trial(
  File "/gantry-runtime/fme/core/generics/lr_tuning.py", line 121, in run_lr_tuning_trial
    baseline_val_loss = validate_stepper(baseline_stepper, baseline_ema)
  File "/gantry-runtime/fme/core/generics/trainer.py", line 378, in _validate_stepper
    return self._validate_stepper_callback(stepper, ema)
  File "/gantry-runtime/fme/ace/train/train.py", line 162, in validate_stepper
    run_validation_loop(
  File "/gantry-runtime/fme/core/generics/validation.py", line 61, in run_validation_loop
    stepped = stepper.train_on_batch(
  File "/gantry-runtime/fme/ace/stepper/single_module.py", line 1595, in train_on_batch
    self._loss_schedule.init_for_epoch(data.epoch)
  File "/gantry-runtime/fme/ace/stepper/loss_schedule.py", line 45, in init_for_epoch
    assert epoch is not None
AssertionError
```

Full log: last failed job [01KTM5CP](https://beaker.org/job/01KTM5CPVXFM3ZNQZ05VY9DN65) (1 failed, 4 preempted across the experiment).

## Cause

The LR-tuning trial's validation callback (`get_validate_stepper_callback` in `fme/ace/train/train.py`) never calls `set_epoch` on the validation data. When a loss schedule with milestones is active, the validation batches therefore carry `epoch=None` into `LossSchedule.init_for_epoch`, which requires an epoch. The bug triggers whenever LR tuning runs a trial while a loss schedule is configured.

`main` is still affected: `validate_stepper` at `fme/ace/train/train.py:152` still takes only `(stepper, ema)`, and `LossSchedule.init_for_epoch` on main raises `EpochNotProvidedError` ("current configuration requires epoch to be provided on BatchData during training") in the same situation.

## Fix

Fixed on `experiment/2026-06-05-aimip-like` by `1ca339ede` ("Fix epoch=None assertion in LR-tuning validation with loss schedule"): thread the current epoch through the `ValidateStepper` protocol and call `data.set_epoch(epoch)` in the callback. That commit (with its tests) needs porting to `main`.

### PR DESCRIPTION
**Pull request by mcgibbon**
LR tuning's validation callback never advanced the validation data to the trial's epoch, so when a loss schedule was active the validation batches carried `epoch=None` and tripped the assertion in `LossSchedule.init_for_epoch` (`EpochNotProvidedError` on main). This crashed the `winds-anomaly-rollout-rs0` residual-winds rollout at epoch 45 ([beaker 01KTG13T](https://beaker.org/ex/01KTG13T8CJHHDDFX187EGKJ6Y)) and recurred on every resume. This threads the trial's epoch through the `ValidateStepper` protocol and calls `data.set_epoch` in the callback, so validation advances to the same epoch as the trial's training forks (and the epoch-based loss schedule selects the matching `n_forward_steps` distribution).

Changes:
- `fme.core.generics.lr_tuning.ValidateStepper` / `run_lr_tuning_trial`: add an `epoch` parameter, forwarded to both the baseline and candidate `validate_stepper` calls
- `fme.core.generics.trainer.Trainer._validate_stepper` / `_maybe_tune_lr`: accept and pass `epoch=self._epochs_trained + 1`
- `fme.ace.train.train.get_validate_stepper_callback` and `fme.coupled.train.train.get_validate_stepper_callback`: accept `epoch` and call `data.set_epoch(epoch)` before validating
- `fme.core.generics.test_lr_tuning` / `test_trainer`: unit-level regression test that the epoch reaches `validate_stepper`, plus the signature updates
- `fme.ace.test_train.test_lr_tuning_with_loss_schedule`: end-to-end regression test that training with both LR tuning and an epoch-based loss schedule (with a milestone) runs without `EpochNotProvidedError`

- [x] Tests added
- [ ] If dependencies changed, "deps only" image rebuilt and "latest_deps_only_image.txt" file updated

Resolves #1275

### CODE REVIEW

**Comment by mcgibbon** [comment]
## Independent pre-review (automated agent, pre-human-review)

Performed by a fresh agent without the authoring context, before this PR is presented as ready. Verdict below; the human reviewer should still review independently.

---

**APPROVE WITH NITS**

This is a correct, well-scoped fix with strong test coverage. Verified end-to-end and at the unit level.

### Correctness
- The root cause is accurate. `LossSchedule.init_for_epoch(None)` raises `EpochNotProvidedError` only when the schedule has milestones (`fme/ace/stepper/loss_schedule.py:32-41`), and the stepper feeds it `data.epoch` (`fme/ace/stepper/single_module.py:1599`). The LR-tuning validation path never called `set_epoch`, so validation batches carried `epoch=None`. The fix threading `data.set_epoch(epoch)` into the callbacks (`fme/ace/train/train.py:157`, `fme/coupled/train/train.py:97`) directly closes this.
- `epoch=self._epochs_trained + 1` is the right value. `_maybe_tune_lr` runs *before* `train_one_epoch` (`trainer.py:449`), the trial trains forks on the upcoming epoch's data, and that same `+1` is already used for the trial's train data (`trainer.py:396`) and for `train_one_epoch`'s `set_epoch`. Validating the trial at the epoch it trained for is consistent with the main path.

### Completeness
- All real `ValidateStepper` implementations and callers are updated: the protocol (`lr_tuning.py:16-22`), both production callbacks, the two `run_lr_tuning_trial` call sites (`lr_tuning.py:127-128`), `Trainer._validate_stepper`/`_maybe_tune_lr`, and both test doubles. No call site missed.

### Tests
- The e2e test genuinely reproduces #1275. `use_schedule=True` builds a `TimeLengthSchedule` with `milestones=[TimeLengthMilestone(epoch=1, ...)]` (`test_train.py:310-320`) — a non-empty milestone list is exactly the `EpochNotProvidedError` trigger. `LRTuningConfig(epochs=Slice())` makes tuning run at epoch 0.
- The unit forwarding test (`test_epoch_forwarded_to_validate_stepper`) asserts `seen_epochs == [4, 4]`, locking in that both baseline and candidate forks get the trial epoch. All 77 tests in `test_lr_tuning.py` + `test_trainer.py` pass.

### Style / Compat
- The conditional scheduler in the helper is necessary and minimal: `lr_tuning` + an explicit scheduler is rejected by `train_config.py:296-300`, and an empty `SchedulerConfig()` has `type=None` so `has_lr_schedule` is False.
- The new required `epoch` parameter is fine — every caller is internal and updated; no backward-compat concern (this is a callback protocol, not checkpoint config).

### Nits (non-blocking)
- The coupled callback (`fme/coupled/train/train.py:97`) gets the identical fix but no dedicated regression test, unlike the ACE path. Given the logic is symmetric and shares the protocol, an explicit coupled e2e test is optional; coupled is covered by the unit-level forwarding test.
- `epoch` could carry a type hint in the test doubles (`test_lr_tuning.py:150,174`) to match the production signatures. Purely cosmetic.

**Review by AnnaKwa — APPROVED** [review]
LGTM
